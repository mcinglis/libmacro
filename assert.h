
// Copyright 2015  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#ifndef LIBMACRO_ASSERT_H
#define LIBMACRO_ASSERT_H


#include <assert.h>

#include <libpp/map.h>          // PP_MAP
#include <libpp/separators.h>   // PP_SEP_SEMICOLON


#ifdef NO_ASSERT
    #define ASSERT( ... )
#else

    // @public
    // Takes a variable number of scalar expressions, and `assert`s each
    // one individually. This provides better assertion error reporting than
    // techniques like `assert( e1 && e2 && ... )`.
    #define ASSERT( ... ) \
        PP_MAP( assert, PP_SEP_SEMICOLON, __VA_ARGS__ )

#endif


#endif


