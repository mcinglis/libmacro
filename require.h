
// Copyright 2015  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#ifndef LIBREQUIRE_REQUIRE_H
#define LIBREQUIRE_REQUIRE_H


#include <libpp/map.h>          // PP_MAP
#include <libpp/separators.h>   // PP_SEP_SEMICOLON
#include <libtypes/types.h>     // noreturn, uint


#ifdef NO_REQUIRE
    #define REQUIRE( ... )
#else

    // @public
    // Checks that all of the given scalar expressions at non-zero. If any
    // are zero, the first zero expression is printed to stderr with the
    // function, file, and line it was made on. Then, execution is
    // `abort()`ed.
    #define REQUIRE( ... ) \
        PP_MAP( REQUIRE1, PP_SEP_SEMICOLON, __VA_ARGS__ )

    #define REQUIRE1( X ) \
        ( ( void )( ( X ) \
                 || ( require_fail( #X, __func__, __FILE__, __LINE__ ), 0 ) ) )

#endif


noreturn
void
require_fail( char const * expr,
              char const * func,
              char const * file,
              uint line );


#endif


