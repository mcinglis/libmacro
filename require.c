
// Copyright 2015  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#include "require.h"

#include <stdio.h>
#include <stdlib.h>

#include "assert.h" // ASSERT


noreturn
void
require_fail( char const * const expr,
              char const * const func,
              char const * const file,
              uint const line )
{
    ASSERT( expr != NULL, file != NULL, func != NULL );

    fprintf( stderr, "%s() (%s:%d) requirement unmet: `%s`\n",
                     func, file, line, expr );
    fflush( NULL );
    abort();
}


