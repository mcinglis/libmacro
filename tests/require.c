
#include "../require.h"


int main( void )
{
    REQUIRE( "these", "requirements", "are", "truthy" );
    REQUIRE( "just", "like this", 21938 * 3 > 12 );
    REQUIRE( 1, 1, 1, 1, 1, 1, 1, 1, 1 );
    // But this is falsy:
    REQUIRE( 123, "falsy incoming" == NULL, 456 );
}


