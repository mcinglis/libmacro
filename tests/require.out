main() (tests/require.c:11) requirement unmet: `"falsy incoming" == ((void *)0)`
